File: packages/socket/src/services/http/result-og.ts
import type { IncomingMessage, ServerResponse } from "http"
import fs from "fs"
import { getResultById } from "@razzoozle/socket/services/config"

// ── per-result Open Graph unfurl (/r/:id) ───────────────────────────────────
// Crawlers fetch /r/:id and read the og:* meta tags. Without this every share
// link previews identically (the static SPA shell). We serve the SAME built
// index.html (so humans get a working app) but rewrite og:title/og:description/
// <title> with this result's winner. Best-effort: a missing result or an
// unreadable file MUST NOT break the page — we always end up serving valid SPA
// HTML (or a 302 to / only when the shell itself is unreadable).
const OG_INDEX_HTML = "/app/web/index.html" // nginx root in the prod image (Dockerfile: COPY web/dist -> /app/web)

const escHtml = (s: string): string =>
  s
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;")

const injectOg = (html: string, title: string, desc: string): string =>
  html
    .replace(
      /(<meta property="og:title" content=")[^"]*(")/i,
      `$1${escHtml(title)}$2`,
    )
    .replace(
      /(<meta property="og:description" content=")[^"]*(")/i,
      `$1${escHtml(desc)}$2`,
    )
    .replace(/(<title>)[^<]*(<\/title>)/i, `$1${escHtml(title)}$2`)

export const handleResultOg = (
  _req: IncomingMessage,
  res: ServerResponse,
  id: string | undefined,
): void => {
  let html: string
  try {
    html = fs.readFileSync(OG_INDEX_HTML, "utf-8")
  } catch {
    // The SPA shell itself is unreadable (misconfigured image): bounce to /.
    res.writeHead(302, { Location: "/" })
    res.end()
    return
  }

  try {
    const result = getResultById(id ?? "")
    const winner = result.players?.[0]
    const subject = result.subject || "Razzoozle"
    const title = winner ? `${subject} — ${winner.username} gewinnt!` : subject
    const desc = winner
      ? `${winner.username} gewinnt mit ${winner.points} Punkten. Spiel selbst auf Razzoozle.`
      : "Endstand auf Razzoozle — spiel selbst mit."
    html = injectOg(html, title, desc)
  } catch {
    // Unknown/corrupt id: serve the SPA shell unchanged (renders not-found) with
    // its default og tags. Never throws past here.
  }

  res.writeHead(200, {
    "Content-Type": "text/html; charset=utf-8",
    "Cache-Control": "no-cache",
  })
  res.end(html)
}

File: rust/server/src/http/result_og.rs
use axum::{
    extract::Path,
    http::StatusCode,
    response::{Html, IntoResponse, Response},
};
use std::path::Path as StdPath;

use super::AppState;
use crate::db::get_result_by_id;

// Path to index.html in prod Docker image (Dockerfile: COPY web/dist -> /app/web)
const OG_INDEX_HTML_PROD: &str = "/app/web/index.html";
// Fallback paths for development
const OG_INDEX_HTML_DEV: &[&str] = &[
    "packages/web/dist/index.html",
    "packages/web/index.html",
    "web/dist/index.html",
];

/// HTML-escape for safe injection into HTML attributes
fn esc_html(s: &str) -> String {
    s.replace('&', "&amp;")
        .replace('<', "&lt;")
        .replace('>', "&gt;")
        .replace('"', "&quot;")
}

/// Inject OG meta tags and title into HTML
/// Uses simple string replacement (same as Node's approach)
fn inject_og(html: &str, title: &str, desc: &str) -> String {
    let title_esc = esc_html(title);
    let desc_esc = esc_html(desc);

    let mut result = html.to_string();

    // Replace og:title content attribute
    if let Ok(re) = regex::Regex::new(r#"(<meta property="og:title" content=")[^"]*(")"#) {
        result = re.replace(&result, format!(r#"$1{}$2"#, title_esc)).to_string();
    }

    // Replace og:description content attribute
    if let Ok(re) = regex::Regex::new(r#"(<meta property="og:description" content=")[^"]*(")"#) {
        result = re.replace(&result, format!(r#"$1{}$2"#, desc_esc)).to_string();
    }

    // Replace title element
    if let Ok(re) = regex::Regex::new(r"(<title>)[^<]*(<\/title>)") {
        result = re.replace(&result, format!(r"$1{}$2", title_esc)).to_string();
    }

    result
}

/// Read index.html from prod/dev paths
fn read_index_html() -> Result<String, Box<dyn std::error::Error>> {
    // Try prod path first
    if StdPath::new(OG_INDEX_HTML_PROD).exists() {
        return Ok(std::fs::read_to_string(OG_INDEX_HTML_PROD)?);
    }

    // Try dev paths
    for dev_path in OG_INDEX_HTML_DEV {
        if StdPath::new(dev_path).exists() {
            return Ok(std::fs::read_to_string(dev_path)?);
        }
    }

    Err("index.html not found in any expected location".into())
}

/// GET /r/:id — Open Graph unfurl for social media share previews
pub async fn handle_result_og(
    Path(id): Path<String>,
    axum::extract::State(state): axum::extract::State<AppState>,
) -> Result<impl IntoResponse, Response> {
    // Read base HTML (302 redirect if SPA shell missing)
    let mut html = match read_index_html() {
        Ok(h) => h,
        Err(_) => {
            return Err((
                StatusCode::FOUND,
                [("Location", "/")],
            ).into_response());
        }
    };

    // Try to load result and inject OG tags
    if let Some(result) = get_result_by_id(&state.db_pool, &id).await {
        if let Some(players) = result.get("players").and_then(|p| p.as_array()) {
            if let Some(first_player) = players.first() {
                let winner_name = first_player
                    .get("username")
                    .and_then(|u| u.as_str())
                    .unwrap_or("Ein Spieler");
                let points = first_player
                    .get("points")
                    .and_then(|p| p.as_i64())
                    .unwrap_or(0);
                let subject = result
                    .get("subject")
                    .and_then(|s| s.as_str())
                    .unwrap_or("Razzoozle");

                let title = format!("{} — {} gewinnt!", subject, winner_name);
                let desc = format!(
                    "{} gewinnt mit {} Punkten. Spiel selbst auf Razzoozle.",
                    winner_name, points
                );

                html = inject_og(&html, &title, &desc);
            }
        }
    }

    Ok((
        StatusCode::OK,
        [("Content-Type", "text/html; charset=utf-8"), ("Cache-Control", "no-cache")],
        Html(html),
    ))
}
